File: prg/src/states.ml
type w = int array

let new_w d =
  Array.make d 0

let copy w =
  Array.copy w
  
let is_null w =
  Array.for_all (fun x -> x = 0) w

let state_space d s =
  Math.c d s

let print_w w =
  Array.iter (fun x -> Printf.printf "%d " x) w;
  Printf.printf "\n"
  
let rec compute_w_i w_set w d s sigma_i i index =
  if i >= d then
    begin
      w_set.(!index) <- Array.copy w;
      index := !index + 1
    end
  else
    begin
      for sigma=sigma_i to (i+1)*s do
        w.(d-i-1) <- sigma;
        compute_w_i w_set w d s sigma (i+1) index
      done
    end
  
let compute_w d s =
  let space = state_space d s in
  let w_set = Array.make space [||] in
  compute_w_i w_set (Array.make d 0) d s 0 0 (ref 0);
  for i=0 to space-1 do
    w_set.(i).(0) <- w_set.(i).(0) - w_set.(i).(1);
    for j=1 to d-2 do
      w_set.(i).(j) <- w_set.(i).(j-1) + w_set.(i).(j) - w_set.(i).(j+1)
    done;
    w_set.(i).(d-1) <- w_set.(i).(d-2) + w_set.(i).(d-1);
  done;
  w_set

let get w i =
  if i = 0 then
    0
  else
    w.(i-1)

let add w i e =
  let d = Array.length w in
  let rec add_aux w i e =
    if i < d then
      begin
        w.(i) <- w.(i) + e;
        add_aux w (i+1) e
      end
    else
      () in
  add_aux w (i-1) e

let inc_time w v =
  let d = Array.length w in
  for i=0 to d-2 do
    w.(i) <- max (w.(i+1)-v) 0;
  done;
  w.(d-1) <- w.(d-2);
  ()
    
let set w x e =
  let d = Array.length w in
  w.(x-1) <- e;
  if (x < d && w.(x) < w.(x-1)) || (x > 1 && w.(x-2) > w.(x-1)) then
    failwith("States.set: not a valid value, this is not a valid remaining function anymore")
  else
    ()

File: prg/src/c++/states.cpp
#include <iostream>
#include "math.h"
#include "states.hpp"

W::W(void): w(NULL), m_d(0){
}

W::W(unsigned int d): m_d(d){
  w = new unsigned int[d];
}

W::W(const W& obj){
  m_d = obj.m_d;
  w = new unsigned int[m_d];
  for (unsigned int i = 0; i < m_d; i++)
    w[i] = obj.w[i];
}

unsigned int& W::operator[](unsigned int i){
  return w[i];
}

void W::operator=(W obj){
  if (m_d != 0)
    delete[] w;
  w = new unsigned int[obj.m_d];
  m_d = obj.m_d;
  for (unsigned int i = 0; i < m_d; i++)
    w[i] = obj.w[i];
}

bool W::operator==(const W& obj) const{
  for (unsigned int i = 0; i < m_d; i++)
    if (w[i] != obj.w[i])
      return false;
  return true;
}

bool W::operator<(const W& obj) const{
  unsigned int d = m_d;
  for (unsigned int i = 0; i < d; i++)
    if (w[i] != obj.w[i])
      return w[i] < obj.w[i];
  return w[d-1] < obj.w[d-1];
}

bool W::is_null(void){
  for (unsigned int i = 0; i < m_d; i++)
    if (w[i] != 0)
      return false;
  return true;
}

void W::add_work(unsigned int delta, unsigned int sigma){
  for (unsigned int i = delta-1; i < m_d; i++)
    w[i] += sigma;
}


void W::print_w(void){
  for (unsigned int i = 0; i < m_d-1; i++)
    std::cout << w[i] << " ";
  std::cout << w[m_d-1] << std::endl;
}

unsigned int W::get(unsigned int i) const {
  if (i == 0)
    return 0;
  else
    return w[i-1];
}

unsigned int W::size(void) const{
  return m_d;
}

void W::add(unsigned int i, unsigned int e){
  unsigned int d = m_d;

  for (unsigned int j = i-1; j < d; j++)
    w[j] += e;
}

void W::inc_time(unsigned int v){
  unsigned int d = m_d;
  for (unsigned int i = 0; i < d-1; i++)
    w[i] = minus(w[i+1], v);
  w[d-1] = w[d-2];
}

void W::set(unsigned int x, unsigned int e){
  unsigned int d = m_d;

  w[x-1] = e;
  if ((x < d && w[x] < w[x-1]) || (x > 1 && w[x-2] > w[x-1])){
    std::cout << "States.set: not a valid value, this is not a valid remaining function anymore" << std::endl;
  }
}

unsigned int state_space(unsigned int d, unsigned int s){
  return c(d,s);
}


void compute_w_i(W* w_set, W w, unsigned int d, unsigned int s, unsigned int sigma_i, unsigned int i, unsigned int &index){
  if (i >= d){
    w_set[index] = w;
    index++;
  }
  else
    for (unsigned int sigma = sigma_i; sigma <= (i+1)*s; sigma++){
      w[d-i-1] = sigma;
      compute_w_i(w_set, w, d, s, sigma, (i+1), index);
    }
}

W* compute_w(unsigned int d, unsigned int s){
  unsigned int space = state_space(d, s), index = 0;
  W* w_set = NULL;
  w_set = new W[space];
  if  (w_set == NULL){
    printf("Error: memory allocation for w_set did not work!\n");
    exit(1);
  }
  
  W w(d);

  compute_w_i(w_set, w, d, s, 0, 0, index);

  for (unsigned int i = 0; i < space; i++){
    w_set[i][0] -= w_set[i][1];
    for (unsigned int j = 1; j < d-1; j++)
      w_set[i][j] += w_set[i][j-1] - w_set[i][j+1];
    w_set[i][d-1] += w_set[i][d-2];
  }

  return w_set;
}


